crates/astria-sequencer/src/app/proposal_fingerprint.rs

```diff
@@ -0,0 +1,192 @@
+    //
+    // Can only be run on an unset fingerprint.
+    pub(crate) fn set_prepared_proposal(
+        &mut self,
+        request: abci::request::PrepareProposal,
+        response: abci::response::PrepareProposal,
+    ) -> Result<()> {
+        use prost::Message as _;
+        use sha2::{
+            Digest as _,
+            Sha256,
+        };
+        use tendermint_proto::v0_38::abci as pb;
+        if self.0 != ProposalFingerprintData::Unset {
+            bail!("ProposalFingerprint already set");
+        }
+
+        let proposed_last_commit = if let Some(local_last_commit) = request.local_last_commit {
+            let vote_info = local_last_commit
+                .votes
+                .into_iter()
+                .map(|vote| abci::types::VoteInfo {
+                    validator: vote.validator,
+                    sig_info: vote.sig_info,
+                })
+                .collect();
+            Some(abci::types::CommitInfo {
+                round: local_last_commit.round,
+                votes: vote_info,
+            })
+        } else {
+            None
+        };
+        let proposal = abci::request::ProcessProposal {
+            hash: Hash::default(),
+            proposed_last_commit,
+            height: request.height,
+            time: request.time,
+            proposer_address: request.proposer_address,
+            next_validators_hash: request.next_validators_hash,
+            misbehavior: request.misbehavior,
+            txs: response.txs,
+        };
+
+        let pb_data = pb::RequestProcessProposal::from(proposal).encode_to_vec();
+        let data: [u8; 32] = Sha256::digest(pb_data).into();
+        self.0 = ProposalFingerprintData::Prepared(data);
+        Ok(())
+    }
+
+    // Given a ProcessProposal request, check the ProcessProposal matches
+    // the fingerprint from prepare_proposal. If in `PreparedProposal` state
+    // will create a partial ProcessProposal message, serialize and hash that
+    // to compare. If it does not match, the status is set to `InvalidChecked`.
+    // The overall status is returned.
+    pub(crate) fn check_if_prepared_proposal(
+        &mut self,
+        proposal: &abci::request::ProcessProposal,
+    ) -> bool {
+        use prost::Message as _;
+        use sha2::{
+            Digest as _,
+            Sha256,
+        };
+        use tendermint_proto::v0_38::abci as pb;
+        match self.0 {
+            ProposalFingerprintData::Unset
+            | ProposalFingerprintData::InvalidCheckedPrepared(_)
+            | ProposalFingerprintData::InvalidCheckedExecutedBlock(..)
+            | ProposalFingerprintData::ExecutedBlock(..) => false,
+            ProposalFingerprintData::PreparedValid(_) => true,
+            ProposalFingerprintData::Prepared(proposal_hash) => {
+                let partial_proposal = abci::request::ProcessProposal {
+                    hash: Hash::default(),
+                    ..proposal.clone()
```

Comment:
don't think we need to clone here?

Reply:
Unfortunately without a clone I believe I would need to take ownership, or break out the fields. Misbehavior can't be copied, lol